### PR TITLE
Feature/generate stubs

### LINKED ISSUE #15: Stub file generation
Generate stub files from the data generated from the tracing. Find a library to assist this.

### PR DESCRIPTION
**Pull request by RecurvedBow**
Closes #15

### CODE REVIEW

File: typegen/strats/stub.py [hunk 1]
```diff
@@ -1,5 +1,117 @@
...
+from mypy import stubgen
+import constants
+from tracing.ptconfig import load_config
+from typegen.strats.inline import InlineGenerator
 
 
-class StubFileGenerator(TypeHintGenerator):
-    ident = "stub"
\ No newline at end of file
+def _get_temp_folder_path(root: pathlib.Path) -> pathlib.Path:
```
Comment (bengsparks) [inline typegen/strats/stub.py:9]: This sort of functionality should be covered by the builtin `tempfile` module, which is very nice to use

**Review by bengsparks — CHANGES_REQUESTED** [review]

File: typegen/strats/stub.py (hunk 1 above)
Comment (RecurvedBow) [reply typegen/strats/stub.py]: Done

File: tests/typegen/strats/test_stub.py [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+
+def test_stub_file_generator_generates_file_with_correct_content_with_union_import():
+    resource_path = pathlib.Path("tests", "resource", "typegen", "sample_file_with_type_hint_unions.py")
+    empty_trace_data = pd.DataFrame(columns=Schema.TraceData)
+
+    empty_trace_data.loc[len(empty_trace_data.index)] = [
+        "",
+        None,
+        None,
+        "",
```
Comment (bengsparks) [inline tests/typegen/strats/test_stub.py:66]: This should be None

File: tests/typegen/strats/test_stub.py [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+
+def test_stub_file_generator_generates_file_with_correct_content_with_custom_union_without_union_import():
+    resource_path = pathlib.Path("tests", "resource", "typegen", "sample_file_with_custom_union.py")
+    empty_trace_data = pd.DataFrame(columns=Schema.TraceData)
+
+    empty_trace_data.loc[len(empty_trace_data.index)] = [
+        "",
+        None,
+        None,
+        "",
```
Comment (bengsparks) [inline tests/typegen/strats/test_stub.py:101]: This should also be None

File: typegen/strats/stub.py [hunk 4]
```diff
@@ -1,5 +1,92 @@
...
+                stub_file_content = file.read()
+
+            stub_cst = cst.parse_module(stub_file_content)
+
+            transformer = _ImportUnionTransformer()
+            stub_cst = stub_cst.visit(transformer)
+
+            super()._store_hinted_ast(path_of_stub_file, stub_cst)
+        finally:
+            # Clears the file content.
```
Comment (bengsparks) [inline typegen/strats/stub.py:91]: Unnecessary; the tempfile is deleted and its contents are lost as soon as the variable goes out of scope /  gets garbage collected

File: typegen/strats/stub.py [hunk 5]
```diff
@@ -1,5 +1,92 @@
...
+
+    def _check_annotation_union(self, node: cst.CSTNode | None) -> None:
+        if self.requires_union_import:
+            return
+        if isinstance(node, cst.Subscript):
+            if isinstance(node.value, cst.Name) and node.value.value == "Union":
+                self.requires_union_import = True
+
+
+class StubFileGenerator(InlineGenerator):
```
Comment (bengsparks) [inline typegen/strats/stub.py:48]: The `StubFileGenerator` is not an inline generator; the types are stored next to the original source code.
Inherit from `TypeHintGenerator` instead, and replace the calls to `super()._store_hinted_ast` appropriately

**Review by bengsparks — CHANGES_REQUESTED** [review]

**Review by bengsparks — CHANGES_REQUESTED** [review]
Check that test data for `CLASS_MEMBER`s use None for the FuncName instead of empty strings

File: typegen/strats/stub.py (hunk 5 above)
Comment (bengsparks) [reply typegen/strats/stub.py]: Refactored and co-authored together

File: tests/typegen/strats/test_stub.py (hunk 2 above)
Comment (RecurvedBow) [reply tests/typegen/strats/test_stub.py:63]: Done

File: tests/typegen/strats/test_stub.py (hunk 3 above)
Comment (RecurvedBow) [reply tests/typegen/strats/test_stub.py:98]: Done

**Comment by RecurvedBow** [comment]
All test data have been updated.

**Review by bengsparks — APPROVED** [review]

**Pull request merged**